astropy/coordinates/builtin_frames/ecliptic_transforms.py

```diff
@@ -87,21 +112,101 @@ def icrs_to_helioecliptic(from_coo, to_frame):
     heliocart = from_coo.cartesian - bary_sun_pos
 
     # now compute the matrix to precess to the right orientation
-    rmat = _ecliptic_rotation_matrix(to_frame.equinox)
+    rmat = _mean_ecliptic_rotation_matrix(to_frame.equinox)
 
     newrepr = heliocart.transform(rmat)
     return to_frame.realize_frame(newrepr)
 
 
 @frame_transform_graph.transform(FunctionTransformWithFiniteDifference,
-                                 HeliocentricTrueEcliptic, ICRS,
+                                 HeliocentricMeanEcliptic, ICRS,
                                  finite_difference_frameattr_name='equinox')
 def helioecliptic_to_icrs(from_coo, to_frame):
     if not u.m.is_equivalent(from_coo.cartesian.x.unit):
         raise UnitsError(_NEED_ORIGIN_HINT.format(from_coo.__class__.__name__))
 
     # first un-precess from ecliptic to ICRS orientation
-    rmat = _ecliptic_rotation_matrix(from_coo.equinox)
+    rmat = _mean_ecliptic_rotation_matrix(from_coo.equinox)
+    intermed_repr = from_coo.cartesian.transform(matrix_transpose(rmat))
+
+    # now offset back to barycentric, which is the correct center for ICRS
+
+    # this goes here to avoid circular import errors
+    from astropy.coordinates.solar_system import get_body_barycentric
+
+    # get barycentric sun coordinate
+    bary_sun_pos = get_body_barycentric('sun', from_coo.obstime)
+
+    newrepr = intermed_repr + bary_sun_pos
+    return to_frame.realize_frame(newrepr)
```

Comment:
Maybe I'm missing something... but isn't this expressable as an `AffineTransform` (http://docs.astropy.org/en/stable/api/astropy.coordinates.AffineTransform.html) since it looks to be a rotation plus an offset? (similar for the other direction)

(The advantage of `AffineTransform` over `FunctionTransformWithFiniteDifference` is that it deals with velocities much better)

Reply:
For the case "rotation then offset" it's easy, but "offset then rotation" is not that straightforward because `AffineTransform` expects the former. In any case, it's a start!

Reply:
Ah, sounds like we should add an option to let us decide the order of operations...

Reply:
See #8601